clientapi/routing/rate_limiting.go

```diff
@@ -38,6 +39,7 @@ func (l *rateLimits) clean() {
 		// empty. If they are then we will close and delete them,
 		// freeing up memory.
 		time.Sleep(time.Second * 30)
+		l.cleanMutex.Lock()
```

Comment:
Why have we sprouted another mutex? Double locking makes me scared of deadlocks.

Reply:
The alternative is that we release the read lock, take a write lock, make the write, release the write lock, retake the read lock, and in that time, the cleaner goroutine might have taken the lock instead.

This just stops the cleaner goroutine from running while requests are in flight.

Reply:
The cleaner was taking a write lock to clean, so I don't see any problem?

Reply:
The cleaner closes the channel if it thinks it's no longer needed, so the alternative was something less ideal, like:
```
l.limits[caller] = make(chan struct{}, l.requestThreshold)

l.limitsMutex.RUnlock()
l.limitsMutex.Lock()
rateLimit = l.limits[caller]
l.limitsMutex.Unlock()

// cleaner could win the lock and run here, closing our channel, and we haven't used it yet

l.limitsMutex.RLock()
```

Reply:
Can we avoid this by assigning it into `l.limits` after we've done our work with it?

Reply:
Another HTTP request could come in in the meantime, which would result in more than one channel for the caller.

Reply:
OK then, SHIPIT